File: src/main/java/ssg/search/collection/es/disp/ESMallCount.java
package ssg.search.collection.es.disp;

import com.google.common.collect.*;
import io.searchbox.core.SearchResult;
import io.searchbox.core.search.aggregation.TermsAggregation;
import org.apache.commons.lang3.StringUtils;
import ssg.search.base.Call;
import ssg.search.base.Info;
import ssg.search.collection.es.ESCollection;
import ssg.search.collection.es.ESQuery;
import ssg.search.function.Groupable;
import ssg.search.parameter.Parameter;
import ssg.search.result.Result;

import java.util.List;
import java.util.Map;

public class ESMallCount extends ESCollection implements Groupable {
	@Override
	public String getCollectionAliasName(Parameter parameter) {
		return "ESMallCount";
	}

	@Override
	public List<String> getQuery(Parameter parameter) {
		Multimap<String, String> queryMap = HashMultimap.create();
		for (ESQuery query : ImmutableSet.of(
				ESQuery.SRCH_PREFIX_MALL
				, ESQuery.DISP_CTG_LST
                , ESQuery.DISP_CTG_SUB_LST
				, ESQuery.BRAND_ID
				, ESQuery.SALESTR_LST_MALL
				, ESQuery.SIZE
				, ESQuery.COLOR
				, ESQuery.DEVICE_CD
				, ESQuery.MBR_CO_TYPE
				, ESQuery.PRC_FILTER
				, ESQuery.SCOM_EXPSR_YN
		)) {
			if (StringUtils.isNotEmpty(query.getValue(parameter))) {
				queryMap.put(query.getField(parameter), query.getValue(parameter));
			}
		}

		List<String> queryList = Lists.newArrayList();
		for (Map.Entry<String, String> e : queryMap.entries()) {
			if (StringUtils.isNotEmpty(e.getValue())) {
				queryList.add(e.getKey() + ":(" + e.getValue() + ")");
			}
		}

		return queryList;
	}

	@Override
	public Call<Info> getGroup() {
		return new Call<Info>() {
			public Info apply(Parameter parameter) {
				return new Info("SITE_NO,SCOM_EXPSR_YN");
			}
		};
	}

	@Override
	public Result getResult(SearchResult searchResult, String name, Parameter parameter, Result result) {
		result.setMallCountMap(getMallCountMap(searchResult, parameter));
		return result;
	}

	public Map<String, String> getMallCountMap(SearchResult searchResult, Parameter parameter) {
		Map<String, String> mallCountMap = Maps.newHashMap();
		String[] fields = getGroup().apply(parameter).getPredicate().split(",");

		for (String f : fields) {
			List<TermsAggregation.Entry> entryList = searchResult.getAggregations().getTermsAggregation(f).getBuckets();
			if (StringUtils.equals(f, "SITE_NO")) {
				for (TermsAggregation.Entry entry : entryList) {
					mallCountMap.put(entry.getKey(), String.valueOf(entry.getCount()));
				}
			} else if (StringUtils.equals(f, "SCOM_EXPSR_YN")) {
				for (TermsAggregation.Entry entry : entryList) {
					if (StringUtils.equals(entry.getKey(), "Y")) {
						mallCountMap.put("6005", String.valueOf(entry.getCount()));
					}

					mallCountMap.put(entry.getKey(), String.valueOf(entry.getCount()));
				}
			}
		}

		return mallCountMap;
	}
}

File: src/main/kotlin/ssg/front/search/api/collection/es/common/EsMallCount.kt
package ssg.front.search.api.collection.es.common

import com.google.common.collect.HashMultimap
import org.elasticsearch.action.search.SearchResponse
import org.elasticsearch.search.aggregations.bucket.terms.ParsedStringTerms
import ssg.front.search.api.collection.es.EsCollection
import ssg.front.search.api.collection.es.EsQuery
import ssg.front.search.api.dto.Parameter
import ssg.front.search.api.dto.result.Result
import ssg.front.search.api.dto.vo.GroupVo
import ssg.front.search.api.function.Groupable


open class EsMallCount : EsCollection(), Groupable {
    override fun getQuery(parameter: Parameter): List<String> {
        val queryMap = HashMultimap.create<String, String>()

        for (query in listOf(
                EsQuery.SRCH_PREFIX_MALL,
                EsQuery.DISP_CTG_LST,
                EsQuery.DISP_CTG_SUB_LST,
                EsQuery.BRAND_ID,
                EsQuery.SALESTR_LST_MALL,
                EsQuery.SIZE,
                EsQuery.COLOR,
                EsQuery.DEVICE_CD,
                EsQuery.MBR_CO_TYPE,
                EsQuery.PRC_FILTER,
                EsQuery.SCOM_EXPSR_YN
        )) {
            if (query.getValue(parameter).isNotEmpty()) {
                queryMap.put(query.getField(parameter), query.getValue(parameter))
            }
        }

        val queryList = mutableListOf<String>()
        for ((key, value) in queryMap.entries()) {
            if (!value.isNullOrEmpty()) {
                queryList.add("$key:($value)")
            }
        }

        return queryList
    }

    override fun groupVo(parameter: Parameter): GroupVo {
        return GroupVo("SITE_NO,SCOM_EXPSR_YN")
    }

    fun getMallCountMap(searchResponse: SearchResponse): HashMap<String, String> {
        val mallCountMap = HashMap<String, String>()
        for (bucket in searchResponse.aggregations.get<ParsedStringTerms>("SITE_NO").buckets) {
            mallCountMap[bucket.key.toString()] = bucket.docCount.toString()
        }

        for (bucket in searchResponse.aggregations.get<ParsedStringTerms>("SCOM_EXPSR_YN").buckets) {
            if (bucket.key.toString() == "Y") {
                mallCountMap["6005"] = bucket.docCount.toString()
            } else {
                mallCountMap[bucket.key.toString()] = bucket.docCount.toString()
            }
        }

        return mallCountMap
    }

    override fun getResult(searchResponse: SearchResponse, name: String, parameter: Parameter, result: Result) {
        result.mallCountMap = getMallCountMap(searchResponse)
    }
}
